torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -544,6 +645,39 @@ bool FunctionSignature::parse(PyObject* args, PyObject* kwargs, PyObject* dst[],
       dst[i++] = args;
       arg_pos = nargs;
       continue;
+    } else if (check_has_torch_function(obj)) {
+      // Collects arguments with __torch_function__ and
+      // in the order in which they should be tried (i.e., skipping redundant types).
+      // by checking for subclasses
+      dst[i++] = obj;
+      for (j = 0; j < num_overloaded_args; j++) {
+        if (Py_TYPE(obj) == Py_TYPE(overloaded_args[j])) {
+          new_class = 0;
+          break;
+        }
+      }
+      if (new_class) {
+        PyObject *method = get_torch_function(obj);
+
+        if (method != nullptr) {
+          int arg_index;
+
+          arg_index = num_overloaded_args;
+
+          for (j = 0; j < num_overloaded_args; j++) {
+            PyObject *other_type;
+            other_type = (PyObject *)Py_TYPE(overloaded_args[j]);
+            if (PyObject_IsInstance(obj, other_type)) {
+              arg_index = j;
+              break;
+            }
+          }
+          pyobject_array_insert(overloaded_args, num_overloaded_args, arg_index, obj);
+          ++num_overloaded_args;
+        }
+      }
+    } else if (param.check(obj)) {
```

Comment:
It would be helpful to explain under what circumstances we can actually fall through to this case.

Reply:
Added comments